spec/app/ics-020-fungible-token-transfer/v1/README.md

```diff
@@ -240,130 +146,142 @@ In plain English, between chains `A` and `B`:
-
-    // send packet using the interface defined in ICS4
-    sequence = handler.sendPacket(
-      getCapability("port"),
-      sourcePort,
-      sourceChannel,
-      timeoutHeight,
-      timeoutTimestamp,
-      json.marshal(data) // json-marshalled bytes of packet data
-    )
-
-    return sequence
+    return true
 }
 ```
 
 `onRecvPacket` is called by the routing module when a packet addressed to this module has been received.
 
 ```typescript
-function onRecvPacket(packet: Packet) {
-  FungibleTokenPacketData data = packet.data
-  assert(data.denom !== "")
-  assert(data.amount > 0)
-  assert(data.sender !== "")
-  assert(data.receiver !== "")
+function onRecvPacket(
+  destChannelId: bytes,
+  sourceChannelId: bytes,
+  sequence: bigEndianUint64, 
+  payload: Payload,
+): (bytes,bool) {
+
+  success,appData=decode(payload.encoding,payload.appData)
+  abortTransactionUnless(success)
+
+  assert(appData.denom !== "")
+  assert(appData.amount > 0)
+  assert(appData.sender !== "")
+  assert(appData.receiver !== "")
 
   // construct default acknowledgement of success
   FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{true, null}
-  prefix = "{packet.sourcePort}/{packet.sourceChannel}/"
+  prefix = "{payload.sourcePort}/{sourceChannelId}/"
   // we are the source if the packets were prefixed by the sending chain
-  source = data.denom.slice(0, len(prefix)) === prefix
+  source = appData.denom.slice(0, len(prefix)) === prefix
   if source {
     // receiver is source chain: unescrow tokens
     // determine escrow account
-    escrowAccount = channelEscrowAddresses[packet.destChannel]
+    escrowAccount = channelEscrowAddresses[destChannelId]
     // unescrow tokens to receiver (assumed to fail if balance insufficient)
-    err = bank.TransferCoins(escrowAccount, data.receiver, data.denom.slice(len(prefix)), data.amount)
+    err = bank.TransferCoins(escrowAccount, appData.receiver, appData.denom.slice(len(prefix)), appData.amount)
     if (err !== nil)
       ack = FungibleTokenPacketAcknowledgement{false, "transfer coins failed"}
   } else {
     prefix = "{packet.destPort}/{packet.destChannel}/"
-    prefixedDenomination = prefix + data.denom
+    prefixedDenomination = prefix + appData.denom
     // sender was source, mint vouchers to receiver (assumed to fail if balance insufficient)
-    err = bank.MintCoins(data.receiver, prefixedDenomination, data.amount)
+    err = bank.MintCoins(appData.receiver, prefixedDenomination, appData.amount)
     if (err !== nil)
       ack = FungibleTokenPacketAcknowledgement{false, "mint coins failed"}
   }
-  return ack
+  return ack,true
 }
 ```
 
 `onAcknowledgePacket` is called by the routing module when a packet sent by this module has been acknowledged.
 
 ```typescript
 function onAcknowledgePacket(
-  packet: Packet,
-  acknowledgement: bytes) {
+  sourceChannelId: bytes,
+  destChannelId: bytes, // Can be nullable
```

Comment:
Why would this be nullable?

Reply:
Yeah, here what I wanted to express is that we call the `onAckPacket` with the parameters specified in the [ics04](https://github.com/cosmos/ibc/blob/84341283c480bcfbd982a58b2d0626255eeaadf1/spec/core/v2/ics-004-channel-and-packet-semantics/README.md#:~:text=OnAcknowledgePacket(packet.channelSourceId%2Cpacket.channelDestId%2Cpacket.sequence%2Cpayload%2Cacknowledgement%2C%20relayer)), but, actually, we don't use all of them. Indeed, while all parameters are necessary for [ics20v2](https://github.com/cosmos/ibc/blob/99eaf71eeae605c521261d54765858b1984de28a/spec/app/ics-020-fungible-token-transfer/README.md#:~:text=function%20onAcknowledgePacket(%0A%20%20sourceChannelId%3A%20bytes%2C%0A%20%20destChannelId%3A%20bytes%2C%20%0A%20%20sequence%3A%20bigEndianUint64%2C%0A%20%20payload%3A%20Payload%2C%20%0A%20%20acknowledgement%3A%20bytes%0A%20%20)%3A%20bool%20%7B) `onAckPacket` logic, `destChannelId` and `sequence` are not used in the ics20v1 `onAckPacket`. 

So the question is how do we want to handle this difference in the interfaces? 
1. Do we want to have a standard way to call applications on the core side? 
2. Do we let the core understand how to call the specific app callback retrieving from the router the callback signature? 

Option 2 should provide more flexibility and seems ok. What do you think?